hack/generator/pkg/codegen/pipeline/implement_convertible_interface.go

```diff
@@ -28,30 +27,38 @@ func ImplementConvertibleInterface(idFactory astmodel.IdentifierFactory) Stage {
 		func(ctx context.Context, state *State) (*State, error) {
 			injector := astmodel.NewInterfaceInjector()
 
-			modifiedTypes := make(astmodel.Types)
-			resources := astmodel.FindResourceTypes(state.Types())
-			for name, def := range resources {
-				resource, ok := astmodel.AsResourceType(def.Type())
-				if !ok {
-					// Skip non-resources (though, they should be filtered out, above)
-					continue
-				}
-
-				if resource.IsStorageVersion() {
-					// The hub storage version doesn't implement Convertible
-					continue
-				}
-
-				convertible := createConvertibleInterfaceImplementation(
-					name, resource, state.ConversionGraph(), idFactory)
-				if convertible.FunctionCount() > 0 {
-					modified, err := injector.Inject(def, convertible)
+			modifiedTypes, err := astmodel.FindResourceTypes(state.Types()).Process(
+				func(def astmodel.TypeDefinition) (*astmodel.TypeDefinition, error) {
+					rsrc, _ := astmodel.AsResourceType(def.Type())
+					hub, ok := state.ConversionGraph().FindHub(def.Name(), state.Types())
+					if !ok {
+						// The hub storage version doesn't implement Convertible
+						return nil, nil
+					}
+
+					// For each PropertyAssignmentFunction, create a conversion function that uses it
+					var conversionFunctions []astmodel.Function
+					for _, fn := range rsrc.Functions() {
+						if propertyAssignmentFn, ok := fn.(*functions.PropertyAssignmentFunction); ok {
+
+							conversionFn := functions.NewResourceConversionFunction(hub, propertyAssignmentFn, idFactory)
+							conversionFunctions = append(conversionFunctions, conversionFn)
+						}
+					}
+
+					// Create the interface implementation and inject into the rsrc
+					impl := astmodel.NewInterfaceImplementation(astmodel.ConvertibleInterface, conversionFunctions...)
```

Comment:
No check on the number of functions anymore - is that because they'll always exist?

Reply:
The conversion functions should exist on every non-hub resource, and hubs are short circuited earlier in the function.